### PR TITLE
9.0 merge upstream 20170607

### PR DESCRIPTION
**Pull request by StefanRijnhart**

### CODE REVIEW

**Comment by StefanRijnhart** [comment]
Conflict:
```
--- a/openerp/modules/migration.py
+++ b/openerp/modules/migration.py
@@@ -69,13 -77,11 +77,20 @@@ class MigrationManager(object)
          stageformat = {
              'pre': '[>%s]',
              'post': '[%s>]',
+             'end': '[$%s]',
          }
++<<<<<<< HEAD
 +        # In openupgrade, remove 'or pkg.installed_version is None'
 +        # We want to always pass in pre and post migration files and use a new
 +        # argument in the migrate decorator (explained in the docstring)
 +        # to decide if we want to do something if a new module is installed
 +        # during the migration.
 +        if not (hasattr(pkg, 'update') or pkg.state == 'to upgrade'):
++=======
+         state = pkg.state if stage in ('pre', 'post') else getattr(pkg, 'load_state', None)
+ 
+         if not (hasattr(pkg, 'update') or state == 'to upgrade') or state == 'to install':
++>>>>>>> odoo/9.0
              return
  
          def convert_version(version):
@@@ -116,12 -119,7 +128,16 @@@
              lst.sort()
              return lst
  
++<<<<<<< HEAD
 +        def mergedict(a, b):
 +            a = a.copy()
 +            a.update(b)
 +            return a
 +
 +        parsed_installed_version = parse_version(pkg.installed_version or '')
++=======
+         parsed_installed_version = parse_version(getattr(pkg, 'load_version', pkg.installed_version) or '')
++>>>>>>> odoo/9.0
          current_version = parse_version(convert_version(pkg.data['version']))
  
          versions = _get_migration_versions(pkg)
```
Resolved as 
```
--- a/openerp/modules/migration.py
+++ b/openerp/modules/migration.py
@@@ -69,13 -77,11 +77,17 @@@ class MigrationManager(object)
          stageformat = {
              'pre': '[>%s]',
              'post': '[%s>]',
+             'end': '[$%s]',
          }
+         state = pkg.state if stage in ('pre', 'post') else getattr(pkg, 'load_state', None)
+ 
 -        if not (hasattr(pkg, 'update') or state == 'to upgrade') or state == 'to install':
 +        # In openupgrade, remove 'or pkg.installed_version is None'
 +        # We want to always pass in pre and post migration files and use a new
 +        # argument in the migrate decorator (explained in the docstring)
 +        # to decide if we want to do something if a new module is installed
 +        # during the migration.
-         if not (hasattr(pkg, 'update') or pkg.state == 'to upgrade'):
++
++        if not (hasattr(pkg, 'update') or state == 'to upgrade'):
              return
  
          def convert_version(version):
@@@ -116,12 -119,7 +125,12 @@@
              lst.sort()
              return lst
  
 +        def mergedict(a, b):
 +            a = a.copy()
 +            a.update(b)
 +            return a
 +
-         parsed_installed_version = parse_version(pkg.installed_version or '')
+         parsed_installed_version = parse_version(getattr(pkg, 'load_version', pkg.installed_version) or '')
          current_version = parse_version(convert_version(pkg.data['version']))
  
          versions = _get_migration_versions(pkg)
```

**Comment by StefanRijnhart** [comment]
Setting to work in progress as this breaks on changes to the import of safe_eval (cf. https://github.com/OCA/OpenUpgrade/pull/692)

**Comment by StefanRijnhart** [comment]
Import of safe_eval fixed in commit https://github.com/OCA/OpenUpgrade/pull/895/commits/49cf90264cd41c09e769d2cda5f66ca10e7e828a

**Review by pedrobaeza — APPROVED** [review]

**Comment by mvaled** [comment]
I'm merging this in my local repo.  I'm going to test it and report back.

**Comment by pedrobaeza** [comment]
@mvaled any news on this?

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by mvaled** [comment]
Well, I have managed to migrate the bulk of a moderate sized DB with this merged.  Modules installed:

```
 account
 account_accountant
 account_analytic_analysis
 account_cancel
 account_chart
 account_voucher
 analytic
 analytic_contract_hr_expense
 auth_crypt
 auth_signup
 base
 base_action_rule
 base_iban
 base_import
 base_phone
 base_setup
 base_vat
 board
 bus
 calendar
 contacts
 crm
 crm_claim
 crm_profiling
 crm_todo
 database_cleanup
 decimal_precision
 document
 edi
 email_template
 fetchmail
 hr
 hr_attendance
 hr_contract
 hr_expense
 hr_holidays
 hr_payroll
 hr_payroll_account
 hr_timesheet
 hr_timesheet_invoice
 hr_timesheet_sheet
 im_chat
 im_odoo_support
 knowledge
 l10n_fr
 l10n_fr_rib
 mail
 mass_editing
 mrp
 note
 payment
 payment_transfer
 procurement
 procurement_jit
 procurement_jit_stock
 product
 project
 project_issue
 project_timesheet
 purchase
 purchase_requisition
 report
 resource
 sale
 sale_crm
 sale_mrp
 sale_service
 sales_team
 sale_stock
 share
 stock
 stock_account
 subscription
 warning
 web
 web_calendar
 web_celery
 web_diagram
 web_gantt
 web_graph
 web_kanban
 web_kanban_gauge
 web_kanban_sparkline
 website
 website_crm
 website_mail
 website_partner
 website_report
 website_sale
 web_tests
 web_view_editor
```

I used the following fork of OpenUpgrade for the migration: https://github.com/merchise-autrement/OpenUpgrade/tree/merchise-predevelop-9.0.  I did have to [tweek](https://github.com/merchise-autrement/OpenUpgrade/commit/e40fd6b955f251ffbcbf03e4153daa7b4035ed76) a couple of things just to avoid issues with our DB.  All others commits are either accepted PRs, pendings PR or modifications we have made to our fork of Odoo for normal operation -- see https://github.com/merchise-autrement/odoo/tree/merchise-develop-8.0.

The only thing I have found is that after migration, some views from removed modules remain in the DB (`edi`, `account_analytic_analysis`) and break the normal operation of Odoo 9 with the migrated DB.  The xml_ids were removed from `ir_module_data`.  There's an [ongoing discussion](https://github.com/OCA/OpenUpgrade/pull/594) about this issue.

In summary, I think this is OK.

**Comment by pedrobaeza** [comment]
Ok, merging.

**Comment by StefanRijnhart** [comment]
@mvaled thanks for the report. Can you confirm that #594 solves the issue of the removed xml_ids?

**Comment by mvaled** [comment]
@StefanRijnhart I'll try #594 on my next round of migration (it takes me a while between each to fix several issues with our addons).  I guess next Monday I will be able to do so.
